tools/rpmuncompress.c

```diff
@@ -151,26 +152,39 @@ static char *doUntar(const char *fn)
     if ((at = getArchiver(fn)) == NULL)
 	goto exit;
 
-    if (dstpath) {
-	    int sr = singleRoot(fn);
+    int needtar = (at->extractable == 0);
 
-	    /* the trick is simple, if the archive has multiple entries,
-	     * just extract it into the specified destination path, otherwise
-	     * strip the first path entry and extract in the destination path
-	     */
-	    rasprintf(&mkdir, "mkdir '%s' ; ", dstpath);
+    if (dstpath) {
+	char * sr = singleRoot(fn);
+
+	/* the trick is simple, if the archive has multiple entries,
+	 * just extract it into the specified destination path, otherwise
+	 * strip the first path entry and extract in the destination path
+	 */
+	rasprintf(&mkdir, "mkdir '%s' ; ", dstpath);
+	if (needtar)
 	    rasprintf(&stripcd, " -C '%s' %s", dstpath, sr ? "--strip-components=1" : "");
+	else {
+	    char * moveup;
+	    if (sr)
+		rasprintf(&moveup, " ; tmp=`mktemp -u -p'%s'` ; mv %s/%s $tmp ; (shopt -s dotglob; mv $tmp/* %s) ; rmdir $tmp ", dstpath, dstpath, sr, dstpath);
+	    rasprintf(&stripcd, "%s'%s'%s", at->dest, dstpath, sr ? moveup : "");
+	    free(moveup);
+	}
+	free(sr);
+    } else {
+	mkdir = strdup("");
+	stripcd = strdup("");
```

Comment:
When in rpm, use rstrdup() aka xstrdup() instead (although in the next weeks we'll need to stop adding more C based stuff in the first place)